message/v1/message.go

```diff
@@ -215,7 +219,18 @@ func (mh *MessageHandler) newMessageFromProto(p peer.ID, pbm *pb.Message) (messa
 		if err != nil {
 			return message.GraphSyncMessage{}, err
 		}
-		responses[id] = message.NewResponse(id, graphsync.ResponseStatusCode(res.Status), exts...)
+		var md metadata.Metadata
+		for _, ext := range exts {
+			if ext.Name == graphsync.ExtensionMetadata {
+				var err error
+				md, err = metadata.DecodeMetadata(ext.Data)
+				if err != nil {
+					log.Warnf("Unable to decode metadata in response for request id %d: %w", id, err)
+				}
+			}
```

Comment:
we should remove the metadata extension from the list after processing it.

Reply:
upon further reflection, i think this code should live in this function https://github.com/ipfs/go-graphsync/blob/e21664f437b58b6a5df05fe2450c081ed63162ac/message/v1/message.go#L283 and you can change the return signature to be ([]graphsync.ExtensionData, metadata.Metadata, err)